### PR TITLE
Remove temporal to kernels_arrow

### LINKED ISSUE #5704: Timestamp and Interval arithmetics
### Is your feature request related to a problem or challenge?

When we write such queries:
1)`SELECT val, ts1 - ts2 AS ts_diff FROM table_a ORDER BY ts2 - ts1`, 
2)`SELECT val, interval1 - interval2 AS interval_diff FROM table_a ORDER BY interval2 - interval1 `, 
3)`SELECT val, ts1 - interval1 AS ts_interval_diff FROM table_a ORDER BY ts1 - interval1 DESC`
we get errors like: 
1)`Plan("'Timestamp(Second, None) - Timestamp(Second, None)' is an unsupported operation. addition/subtraction on dates/timestamps only supported with interval types")`,
2)`Plan("'Interval(DayTime) - Interval(DayTime)' can't be evaluated because there isn't a common type to coerce the types to")`,
3)`Execution("Columnar execution is not yet supported for DateIntervalExpr")`

These operations are applicable for scalar values; however, columnar operations are not supported yet.

### Describe the solution you'd like

In planner.rs, binary.rs and datetime.rs, there are some match expressions that need to be extended to handle these cases. Considering the behavior in scalar value operations, required arithmetic operations can be implemented.

### Describe alternatives you've considered

-

### Additional context

-

### PR DESCRIPTION
**Pull request by berkaysynnada**
# Which issue does this PR close?

Closes #5704.

# Rationale for this change

Temporal arithmetic code will be moved to arrow-rs eventually. For this purpose, this PR is an intermediate step that facilitates the process (the same pattern used for other removals).

# What changes are included in this PR?

In datetime.rs, the `evaluate()` function calls the binary.rs function `resolve_temporal_op()` and `resolve_temporal_op_scalar()`, according to type of the second value, Array or Scalar. It can access the **add/sub_dyn...()** functions in kernels_arrow.rs (the functions that will be moved to arrow-rs deployed here until arrow-rs changes happen)

# Are these changes tested?

Yes, with existing tests.

# Are there any user-facing changes?

### CODE REVIEW

**Comment by berkaysynnada** [comment]
@alamb I hope it was in line with your plan to move the codes. If there are places you think would be better, I'd love to hear.

**Review by alamb — APPROVED** [review]
Thank you @berkaysynnada  -- this is definitely in the right direction. 

Eventually my hope is to use the `add_dyn` kernels directly. This feature was added in https://github.com/apache/arrow-rs/pull/4038 by @Weijun-H which will be in arrow 38 which should be landing in DataFusion sometime next week

**Comment by alamb** [comment]
Screenshot 2023-04-21 at 4 36 43 PM

@berkaysynnada  can you please resolve the conflicts with this PR? I don't really understand the issue, but I suspect it will be straightforward to fix

**Review by jackwener — APPROVED** [review]
Thanks @berkaysynnada

File: datafusion/physical-expr/src/expressions/datetime.rs [hunk 1]
```diff
@@ -151,13 +143,18 @@ impl PhysicalExpr for DateTimeIntervalExpr {
...
             }
-            (ColumnarValue::Array(array_lhs), ColumnarValue::Scalar(operand_rhs)) => {
-                evaluate_temporal_array(array_lhs, sign, &operand_rhs)
-            }
-
-            (ColumnarValue::Array(array_lhs), ColumnarValue::Array(array_rhs)) => {
-                evaluate_temporal_arrays(&array_lhs, sign, &array_rhs)
+            // This function evaluates temporal array vs scalar operations, such as timestamp - timestamp,
+            // interval + interval, timestamp + interval, and interval + timestamp. It takes two arrays as input
+            // and an integer sign representing the operation (+1 for addition and -1 for subtraction).
```
Comment (viirya) [inline datafusion/physical-expr/src/expressions/datetime.rs:148]:
```suggestion
            // interval + interval, timestamp + interval, and interval + timestamp. It takes one array and one scalar as input
            // and an integer sign representing the operation (+1 for addition and -1 for subtraction).
```

**Pull request merged**